src/main/java/redis/clients/jedis/JedisClusterInfoCache.java

```diff
@@ -79,36 +80,41 @@ public void discoverClusterNodesAndSlots(Jedis jedis) {
   }
 
   public void discoverClusterSlots(Jedis jedis) {
-    w.lock();
+    //If rediscovering is already in process - no need to start one more same rediscovering, just return
+    if (!rediscovering) {
+      w.lock();
+      rediscovering = true;
```

Comment:
It shouldn't help. We set it on true in lock and to false before leaving lock, so if we got write lock - it's always false in this var.